src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs

```diff
@@ -13,42 +13,80 @@ namespace Microsoft.Interop
+        /// Source generation attempts to reconcile the notion of blittability
+        /// with the C# notion of an "unmanaged" type. This can be accomplished through
+        /// the use of DisableRuntimeMarshallingAttribute.
+        /// </remarks>
+        /// <param name="type">The type to check.</param>
+        /// <returns>Returns true if considered blittable, otherwise false.</returns>
+        public static bool IsConsideredBlittable(this ITypeSymbol type)
+        {
+            unsafe
+            {
+                return IsBlittableWorker(type, ImmutableHashSet.Create<ITypeSymbol>(SymbolEqualityComparer.Default), &IsConsideredBlittableWorker);
+            }
 
-        private static bool HasOnlyBlittableFields(this ITypeSymbol type, ImmutableHashSet<ITypeSymbol> seenTypes)
+            static bool IsConsideredBlittableWorker(ITypeSymbol t, ImmutableHashSet<ITypeSymbol> seenTypes)
+            {
+                return t.IsUnmanagedType;
+            }
+        }
+
+        /// <summary>
+        /// Is the type strictly blittable.
+        /// </summary>
+        /// <remarks>
+        /// Source generation uses a heavily restricted defintion for strictly blittable.
+        /// The definition is based on the built-in marshallers blittable definition but further
+        /// restricts the definition to require only uses primitive types (not including char or bool)
+        /// and do types defined in the source being compiled.
+        /// </remarks>
+        /// <param name="type">The type to check.</param>
+        /// <returns>Returns true if strictly blittable, otherwise false.</returns>
+        public static bool IsStrictlyBlittable(this ITypeSymbol type)
         {
-            if (seenTypes.Contains(type))
+            unsafe
             {
-                // A recursive struct type is illegal in C#, but source generators run before that is detected,
-                // so we check here to avoid a stack overflow.
-                return false;
+                return IsBlittableWorker(type, ImmutableHashSet.Create<ITypeSymbol>(SymbolEqualityComparer.Default), &IsStrictlyBlittableWorker);
             }
 
-            foreach (IFieldSymbol field in type.GetMembers().OfType<IFieldSymbol>())
+            static unsafe bool IsStrictlyBlittableWorker(ITypeSymbol t, ImmutableHashSet<ITypeSymbol> seenTypes)
             {
-                if (!field.IsStatic)
+                if (t.SpecialType is not SpecialType.None)
+                {
+                    return t.SpecialType.IsAlwaysBlittable();
+                }
+                else if (t.IsValueType)
                 {
-                    if (!IsConsideredBlittable(field.Type, seenTypes.Add(type)))
+                    // If the containing assembly for the type is backed by metadata (non-null),
+                    // then the type is not internal and therefore coming from a reference assembly
+                    // that we can not confirm is strictly blittable.
```

Comment:
Can you add a test for this that checks we don't consider the referenced type blittable?

Reply:
Example test that creates a metadata reference from source and then a compilation that references it: https://github.com/dotnet/runtime/blob/d07dee0dabec8bc81d0d0ed042649c4c20c811d9/src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexGeneratorParserTests.cs#L732-L750